fix and improve content on reference pages of table elements (#27958)
mdn/content: files/en-us/web/html/element/table/index.md

{{HTMLSidebar}}

The **`<table>`** [HTML](/en-US/docs/Web/HTML) element represents tabular data—that is, information presented in a two-dimensional table comprised of rows and columns of cells containing data.

{{EmbedInteractiveExample("pages/tabbed/table.html","tabbed-taller")}}

## Attributes

This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attributes).

### Deprecated attributes

The following attributes are deprecated and should not be used. They are documented below for reference when updating existing code and for historical interest only.

- `align` {{deprecated_inline}}

  - : This {{Glossary("enumerated", "enumerated")}} attribute indicates how the table must be aligned inside the containing document. Possible values are:

>     - `left`: the table is displayed on the left side of the document;

Author: Oh, you're right of course ;) The wording in the old spec says "with respect to the document", which I understand how it's meant now...

- `width` {{deprecated_inline}}

  - : This attribute specifies the width of the table.

    > **Note:** Do not use this attribute as it is deprecated (and not supported) in the latest standard: use the CSS {{cssxref("width")}} property instead.
    >
    > While no HTML specification includes `height` as a `<table>` attribute, some browsers support a non-standard interpretation of `height`. The unitless value sets a minimum absolute height in pixels. If set as a percent value, the minimum table height will be relative to the height of the parent container.

## Examples

You'll find some examples below, but for more examples and an in-depth tutorial, see the [HTML tables](/en-US/docs/Learn/HTML/Tables) series in our [Learn web development](/en-US/docs/Learn) area, where you'll learn how to use the table elements and their attributes to get just the right structuring for your tabular data. Visit the [Styling tables](/en-US/docs/Learn/CSS/Building_blocks/Styling_tables) guide in our [Learn web development](/en-US/docs/Learn) area to learn about table styling and some common and useful styling techniques.

> Since the structure of a `<table>` involves the use of several table-related HTML elements along with various associated attributes, the following examples are intended to provide a simplified explanation that covers the basics and common standards. Additional and more detailed information can be found on the corresponding linked pages.

Reviewer: explaining that each all the examples are building on each other, with examples being an h2, each individual example within the larger example being an h3, the HTML, CSS headings being h4, and no h5s.

The goal of this more detailed example of a basic table is to create an accessible yet visually appealing table that displays a list of people's names along with information about their membership in a club.

#### Basic table setup

Because of the way a `<table>` must be structured in HTML, the {{Glossary("markup", "markup")}} can quickly grow into numerous lines of code. For this reason, it's important to have a clear idea of what the table should look like in the end. If the HTML markup of the table is prepared in a good and logical way, it's not only {{Glossary("semantics", "semantically")}} correct and helpful for assistive technologies, but it also makes it easier to style the individual parts of the table later in the CSS.

##### HTML

First, the very basic HTML structure of the table is created. There are no table content groups such as a head or foot, no cells span multiple rows or columns, no relationships between cells are explicitly defined, and no table caption is provided—for now. Basic styling is then added to create lines around the components of the table to make the data structure clearer.

> The table comprises four rows ({{HTMLElement("tr")}} elements), including a header row ({{HTMLElement("th")}} elements within the first row). Each row contains four columns, including a header column ({{HTMLElement("th")}} elements as the first child elements of the rows) and three data columns ({{HTMLElement("td")}} elements). As table sectioning elements are not used, the browser automatically defines the content group structure, i.e., all rows are within the body of the table ({{HTMLElement("tbody")}} element).

Reviewer: some of the content was moved to the intro. This should just be the stuff for the 2nd example.

```html
<table>
  <tr>
    <th>Name</th>
    <th>ID</th>
    <th>Member Since</th>
    <th>Balance</th>
  </tr>
  <tr>
    <th>Margaret Nguyen</th>
    <td>427311</td>
    <td><time datetime="2010-06-03">June 3, 2010</time></td>
    <td>0.00</td>
  </tr>
  <tr>
    <th>Edvard Galinski</th>
    <td>533175</td>
    <td><time datetime="2011-01-13">January 13, 2011</time></td>
    <td>37.00</td>
  </tr>
  <tr>
    <th>Hoshi Nakamura</th>
    <td>601942</td>
    <td><time datetime="2012-07-23">July 23, 2012</time></td>
    <td>15.00</td>
  </tr>
</table>
```

##### CSS

This CSS simply adds a solid border around the `<table>` and around each of the table's cells, including those specified with both {{HTMLElement("th")}} and {{HTMLElement("td")}} elements. That way, both header and data cells are demarcated.

```css
table {
  border: 2px solid rgb(140, 140, 140);
}

th,
td {
  border: 1px solid rgb(160, 160, 160);
}
```

> ##### Result

Reviewer: change all the h5 to h4, and change the embededliveexample links to the renamed example titles.

## Examples

The examples below include tables of progressively increasing complexity. For additional examples, including an in-depth tutorial, see the [HTML tables](/en-US/docs/Learn/HTML/Tables) series in the [Learn web development](/en-US/docs/Learn) area, where you'll learn how to use the table elements and their attributes to correctly structure your tabular data. A [Styling tables](/en-US/docs/Learn/CSS/Building_blocks/Styling_tables) guide provides table styling information, including common, useful techniques.

Since the structure of a `<table>` involves the use of several table-related HTML elements along with various associated attributes, the following examples are intended to provide a simplified explanation that covers the basics and common standards. Additional and more detailed information can be found on the corresponding linked pages.

> These table examples demonstrate how a complete and {{Glossary("accessibility", "accessible")}} table is structured with HTML and styled with [CSS](/en-US/docs/Web/CSS).

Author: ```suggestion
These table examples demonstrate how to create an {{Glossary("accessibility", "accessible")}} table that is structured with HTML and styled with [CSS](/en-US/docs/Web/CSS).
```

#### Result

There is no custom [CSS](/en-US/docs/Web/CSS) or [user stylesheet](/en-US/docs/Web/CSS/Cascade#author_stylesheets) for the table. The styling results purely from the [user-agent stylesheet](/en-US/docs/Web/CSS/Cascade#user-agent_stylesheets).

{{EmbedLiveSample('Basic_table', 650, 80)}}

### Expanded table with header cells

This example extends the [basic table](#basic_table), extending the content and adding basic CSS styles.

#### HTML

> The table comprises four rows ({{HTMLElement("tr")}} elements) now, including a header row ({{HTMLElement("th")}} elements within the first row). Each row contains four columns, including a header column ({{HTMLElement("th")}} elements as the first child elements of the rows) and three data columns ({{HTMLElement("td")}} elements). As table sectioning elements are not used, the browser automatically defines the content group structure, i.e., all rows are wrapped within the body of the table ({{HTMLElement("tbody")}} element).

Author: ```suggestion
The table comprises four rows ({{HTMLElement("tr")}} elements) now, with four columns each. The first row is a row of header cells (The first row contains only {{HTMLElement("th")}} elements). Subsequent rows include a header column ({{HTMLElement("th")}} elements as the first child elements of each row) and three data columns ({{HTMLElement("td")}} elements). As table sectioning elements are not used, the browser automatically defines the content group structure, i.e., all rows are wrapped within the body of the table of an implicit {{HTMLElement("tbody")}} element.
```

> The table comprises four rows ({{HTMLElement("tr")}} elements) now, including a header row ({{HTMLElement("th")}} elements within the first row). Each row contains four columns, including a header column ({{HTMLElement("th")}} elements as the first child elements of the rows) and three data columns ({{HTMLElement("td")}} elements). As table sectioning elements are not used, the browser automatically defines the content group structure, i.e., all rows are wrapped within the body of the table ({{HTMLElement("tbody")}} element).

Author: corrected "column" to "columns" in the first sentence

## Examples

### Basic data table

```html
<table>
  <caption>Alumni Directory</caption>
  <thead>
    <tr>
      <th>First Name</th>
      <th>Last Name</th>
      <th>Year</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Maria</td>
      <td>Sanchez</td>
      <td>2014</td>
    </tr>
    <tr>
      <td>Michael</td>
      <td>Johnson</td>
      <td>2014</td>
    </tr>

>   </tbody>

Reviewer: The original example didn't show best practices. We also want to avoid "simple. And we should iterated that tables are for tabular data

  <tr>
    <th>Name</th>
    <th>ID</th>
    <th>Member Since</th>
    <th>Balance</th>
  </tr>
  <tr>
    <th>Margaret Nguyen</th>
    <td>427311</td>
    <td><time datetime="2010-06-03">June 3, 2010</time></td>
    <td>0.00</td>
  </tr>
  <tr>
    <th>Edvard Galinski</th>
    <td>533175</td>
    <td><time datetime="2011-01-13">January 13, 2011</time></td>
    <td>37.00</td>
  </tr>
  <tr>
    <th>Hoshi Nakamura</th>
    <td>601942</td>
    <td><time datetime="2012-07-23">July 23, 2012</time></td>
    <td>15.00</td>
  </tr>
</table>
```

##### CSS

This CSS just adds a solid black border around the `<table>` and around each of the table's cells, including those specified using both {{HTMLElement("th")}} and {{HTMLElement("td")}}. That way, both header and data cells are demarcated.

```css
table {
  border: 1px solid black;
}

th,
td {
  border: 1px solid black;
}
```

##### Result

{{EmbedLiveSample("Basic_example", 500, 125)}}

#### Row and column spanning

This table builds on the previous example, adding a column for a membership end date. An additional row is also added to include a "Membership Dates" header as a heading for the "Joined" and "Canceled" headers.

##### HTML

> The HTML for the data within the table is similar to the previous example, except for the addition of the new column in each data row. The main changes are in the header. The creation of the super-heading involves adding row and column spans to `<th>` elements to merge elements from multiple rows and columns.  The header spans two rows to create heading cells that are correctly lined up with the content in the body of the table.

Reviewer: ```suggestion
The HTML for the data within the table is similar to the previous example, except for the addition of the new column in each data row. The main changes are in the header. The creation of the super-heading involves adding row and column spans to `<th>` elements to merge elements from multiple rows and columns. The header spans two rows to create heading cells that are correctly lined up with the content in the body of the table.
```

    </tr>
  </thead>
  <tbody>
    <tr>
      <th scope="row">Margaret Nguyen</th>
      <td>427311</td>
      <td><time datetime="2010-06-03">June 3, 2010</time></td>
      <td>n/a</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th scope="row">Edvard Galinski</th>
      <td>533175</td>
      <td><time datetime="2011-01-13">January 13, 2011</time></td>
      <td><time datetime="2017-04-08">April 8, 2017</time></td>
      <td>37.00</td>
    </tr>
    <tr>
      <th scope="row">Hoshi Nakamura</th>
      <td>601942</td>
      <td><time datetime="2012-07-23">July 23, 2012</time></td>
      <td>n/a</td>
      <td>15.00</td>
    </tr>
  </tbody>
</table>
```

Wrapping the rows in {{HTMLElement("thead")}} and {{HTMLElement("tbody")}} as additional semantic markup specifies table content groups.
Also note the use of the `scope` attribute on the {{HTMLElement("th")}} elements within the body content of the table. This defines the cells to which the header element refers, which in our example are all the cells in the `row`.

Once again, we haven't touched the CSS.

```css hidden
table {
  border: 1px solid black;
}

th,
td {
  border: 1px solid black;
}
```

##### Result

> The output is entirely unchanged, despite the addition of useful contextual information under the hood:

Reviewer: ```suggestion
Visually, the output is unchanged.  The additions provide useful contextual information for assistive technologies, including screen readers and search engines:
```

##### HTML

The HTML is similar to the previous example's, except for the addition of the new column in each data row, and the changes to the header. Those changes make the HTML look like this:

```html
<table>
  <tr>
    <th rowspan="2">Name</th>
    <th rowspan="2">ID</th>
    <th colspan="2">Membership Dates</th>
    <th rowspan="2">Balance</th>
  </tr>
  <tr>
    <th>Joined</th>
    <th>Canceled</th>
  </tr>
  <tr>
    <th>Margaret Nguyen</th>
    <td>427311</td>
    <td><time datetime="2010-06-03">June 3, 2010</time></td>
    <td>n/a</td>
    <td>0.00</td>
  </tr>
  <tr>
    <th>Edvard Galinski</th>
    <td>533175</td>
    <td><time datetime="2011-01-13">January 13, 2011</time></td>
    <td><time datetime="2017-04-08">April 8, 2017</time></td>
    <td>37.00</td>
  </tr>
  <tr>
    <th>Hoshi Nakamura</th>
    <td>601942</td>
    <td><time datetime="2012-07-23">July 23, 2012</time></td>
    <td>n/a</td>
    <td>15.00</td>
  </tr>
</table>
```

> The differences that matter here—for the purposes of discussing row and column spans—are in the first few lines of the code above. Note the use of `rowspan` to make the "Name", "ID", and "Balance" headers occupy two rows instead of just one, and the use of `colspan` to make the "Membership Dates" header cell span across two columns.

Reviewer: ```suggestion
Because the dates area spans two rows, the other headers also need to span two rows. We include the `rowspan` attribute on the other  {{HTMLElement("th")}} elements  to make the "Name", "ID", and "Balance" headers occupy two rows. We use the `colspan` attribute to make the "Membership Dates" header cell span the two date columns.
```

Reviewer: These example were indeed on the wrong page. However, they should really be on the `th` and `td` page, as these attributes are supported there, or a tutorial or how to, not the table page or the tr page

> The CSS is unchanged from before.
> 

Reviewer: ```suggestion
```
I don't think we need to mention this since we are saying we're continuing the previous examples.

The head section now has two rows, one with the headers ({{HTMLElement("th")}} elements) "Name", "ID", "Membership Dates", and "Balance", and the other with the "Joined" and "Canceled", which are subheaders of the "Membership Dates" header. This is accomplished by:

- The first row's "Name", "ID", and "Balance" header cells span both table header rows by using the [`rowspan`](/en-US/docs/Web/HTML/Element/th#rowspan) attribute, making them each two rows tall.
- The first row's "Membership Dates" header cell spans two columns using the [`colspan`](/en-US/docs/Web/HTML/Element/th#colspan) attribute, causing it to be two columns wide.
- The second row contains only the two header cells "Joined" and "Canceled" because the other three columns are merged with the cells in the first row that span two rows. The two header cells are correctly positioned under the "Membership Dates" header.

### Table caption and column summary

It's a common and advisable practice to provide a summary for the table's content, allowing users to quickly determine the table's relevance. Furthermore, the "Balance" column is summarized by displaying the sum of the balances of the individual members.

#### HTML

> A table summary is added by using a table [caption](#captions) ({{HTMLElement("caption")}} element) as the first child element of the `<table>`.

Reviewer: it's been merged
